MultiOffRamp - family-agnostic messages

elatoskinas opened the pull request:
## Motivation
The Multi-OffRamp can be converted to be family agnostic by using more generic `Any2EVM` structures and breaking the reliance on abi hashing

## Solution
* Remove `metadataHash` and use an implicit metadata hash composed of `(prefix, srcChainSelector, destChainSelector, onRampAddress)`
* Convert `EVM2EVMMessage` to `Any2EVMRampMessage`, which uses a generic `bytes sender`
* Decouple `messageId` from `hash(message)`
* Out-of-scope: EVM2Any OnRamp changes

Sample on why hashes between Any2X and X2Any are no longer equivalent:
```
Solana:
  - messageId: hash(Sol2AnyMessage msg)
  - (Sol2Any message could contain some extra fields that are SOL-specific)

Off-chain:
  - Sol2AnyMessage -> Any2AnyMessage -> Any2EVMMessage
  - Route SOL -> Ethereum

Ethereum:
  - messageId: hash(Any2EVMMessage msg)
  - (Any2EVMMessage will contain some EVM-specific fields, but not the SOL specific fields)

hash(Sol2AnyMessage) != hash(Any2EVMMessage)
```

elatoskinas committed:
feat: make offramp metadataHash configurable with uniqueness check

elatoskinas committed:
Merge branch 'ccip-develop' into feat/any2evm-configurable-metadata-hash

elatoskinas committed:
fix: metadataHash-based tests

elatoskinas committed:
feat: use onRamp instead of metadataHash in offramp

elatoskinas committed:
feat: replace EVM2EVMMessage with Any2EVMRampMessage

elatoskinas committed:
feat: update message hashing function

elatoskinas committed:
fix: implicit metadata hash order

elatoskinas commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -200,4 +240,26 @@ library Internal {
   }
+
+  /// @notice Family-agnostic header for OnRamp & OffRamp messages
+  // TODO: revisit if destChainSelector is required (likely sufficient to have it implicitly in the commit roots)
```
For context, `destChainSelector` is already in the CCIPSendRequested event, so for the off-chain part we are covererd.

On-chain, we could commit only the roots that match the `destChainSelector` (need another validation in the commit report flow?), so we could exclude it from the message if we wanted to optimize.

No strong opinion here though since having the `destChainSelector` in the message gives a complete set of information just by looking at the header

elatoskinas replied:
Including the dst chain selector is about a ~0.08KB contract size increase, and a ~420 gas increase as opposed to not including it.

I think it's worth it to keep it in the header for easier traceability

elatoskinas commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -200,4 +240,26 @@ library Internal {
+    bytes data; // arbitrary data payload supplied by the message sender
+    address receiver; // receiver address on the destination chain
+    uint256 gasLimit; // user supplied maximum gas amount available for dest chain execution
+    // TODO: revisit collapsing tokenAmounts + sourceTokenData into one struct array
```
Too big of a refactor for this PR, but an alternative could be:

```
struct RampTokenAmounts {
  address token;
  uint256 amount;
  bytes sourceTokenData;
}
```

connorwstein replied:
Thoughts on on/off specific Any2EVMTokenTransfer[] and EVM2AnyTokenTransfer[] like in the proposal? My goal there was to remove SourceTokenData struct and improve readability/consistency by avoid the parallel slices of tokenAmounts + sourceTokenData. Also recognizing that the token transfers are now analogous to small messages where sender=sourcePool, receiver=destToken (look up to destPool on dest) and data=destPoolData. I think the original reason for Internal.SourceTokenData was to avoid breaking the signature of CCIPSendRequested but we're ok to do that now

elatoskinas replied:
Ah, good point, if there are fields that are unused between the On/OffRamp, then it definitely makes sense. I'll make it in the next PR

Do we lose any flexibility on not using bytes for the entire source token data? I'm guessing no since these structs are decoded at the OffRamp/OnRamp level?

elatoskinas replied:
Discussed offline, we can collapse this to:

```solidity
struct Any2AnyTokenAmount {
  bytes sourcePoolAddress;
  bytes destTokenAddress;
  bytes extraData;
  uint256 amount;
}
```

elatoskinas commented on contracts/src/v0.8/ccip/offRamp/EVM2EVMMultiOffRamp.sol:
```diff
@@ -782,23 +779,23 @@ contract EVM2EVMMultiOffRamp is IAny2EVMMultiOffRamp, ITypeAndVersion, MultiOCR3
-          _metadataHash(sourceChainSelector, sourceConfigUpdate.onRamp, Internal.EVM_2_EVM_MESSAGE_HASH);
-        currentConfig.onRamp = sourceConfigUpdate.onRamp;
+      if (currentOnRamp.length == 0) {
+        // TODO: is this check sufficient / is an all-0 check needed?
```
Thoughts on this? Do we need a more complex validation that checks that at least 1 byte is non-zero? I'm thinking it would add much more complexity and contract size, maybe we could do this validation off-chain instead

connorwstein replied:
yeah I think this is fine. If the configured onramp is junk then messages will be unexecutable until we configure the correct one

elatoskinas replied:
This validation adds about ~6.4k gas to `applySourceChainConfigUpdates` in the median case, and a 0.07KB contract size increase.

```
  function validateNonZeroBytesAddress(bytes memory data) public pure {
      uint256 length = data.length;
      uint256 chunkCount = length / 32;

      // Iterate over the `bytes` array in chunks of 32 bytes
      for (uint256 i = 0; i < chunkCount; i++) {
          bytes32 chunk;
          assembly {
              chunk := mload(add(data, add(32, mul(i, 32))))
          }
          if (chunk != 0) {
              return;
          }
      }

      // Check any remaining bytes that do not fit into a `bytes32` chunk
      for (uint256 i = chunkCount * 32; i < length; i++) {
          if (data[i] != 0) {
              return;
          }
      }

      revert ZeroAddressNotAllowed();
  }
```

elatoskinas replied:
I recon this should be delegated to off-chain validations

elatoskinas replied:
> until we configure the correct one

We are disallowing on ramp updates though (since we assume onramp change = offramp upgrade), so a misconfig would force us into a re-deployment

connorwstein replied:
I don't mean code change, I mean just like we accidentally configure say 0xa as the onramp when the real one is 0x123. Messages just will fail to execute, then we adjust offramp to point to 0x123 and they will succeed

elatoskinas replied:
> then we adjust offramp to point to 0x123 and they will succeed

Yeah so this is not possible because of this check:

```solidity
      if (currentOnRamp.length == 0) {
         // new chain selector
      } else if (
        keccak256(currentOnRamp) != keccak256(newOnRamp)
      ) {
        revert InvalidStaticConfig(sourceChainSelector);
      }
```


The offramp adjustment would be a complete redeployment of the offramp

connorwstein replied:
ah and that check is there because we want to prevent minSeqNr overwrites?

elatoskinas replied:
Yeah if we allowed overwriting the OnRamp, it would have some implications:
* `chain => seqNumbers => executionStates` would still be kept even though the OnRamp has changed (maybe not an issue since it's off-ramp specific)
* The same message could accidentally be included in the OffRamp and be replayed (since the `_hash` changes due to the different OnRamp, i.e. metadataHash)
* Committed roots are not cleared, so we would have different commit roots with different metadata hashes

I was always going under the assumption that the metadataHash must remain immutable for each OffRamp, to preserve the OnRamp <-> OffRamp dependency

elatoskinas commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -200,4 +240,26 @@ library Internal {
     Execution
   }
+
+  /// @notice Family-agnostic header for OnRamp & OffRamp messages
```
Should also revisit if the header is worth the added gas & contract size increase due to the `message.header` access

elatoskinas replied:
Insignificant savings - <400 gas for execute, and <0.03KB contract size

elatoskinas commented on contracts/src/v0.8/ccip/offRamp/EVM2EVMMultiOffRamp.sol:
```diff
@@ -377,14 +371,16 @@ contract EVM2EVMMultiOffRamp is IAny2EVMMultiOffRamp, ITypeAndVersion, MultiOCR3
-      hashedLeaves[i] = Internal._hash(message, sourceChainConfig.metadataHash);
+      // TODO: verify message.destChainSelector == config.destChainSelector
+      hashedLeaves[i] = Internal._hash(message, sourceChainConfig.onRamp);
+      // TODO: revisit this - is messageID independent of the leaf hash?
```
Is this check still needed in a family-agnostic ramp world? Would the `messageId` be in a potentially different encoding scheme coming from the source chain, meaning we do not need this check anymore?

@connorwstein

elatoskinas replied:
Yeah I see from the spec the intention is that it is indeed not the hash. Updating this

elatoskinas replied:
Removed: https://github.com/smartcontractkit/ccip/pull/1117/commits/7e3a8a0b99c284e2c71c724425aced6942f59ea7

RayXpub commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -158,6 +173,31 @@ library Internal {
+        ),
+        keccak256(
+          abi.encode(
+            original.sender, original.receiver, original.header.sequenceNumber, original.gasLimit, original.header.nonce
```
q: why have `original.feeToken` and `original.feeTokenAmount` been removed?

elatoskinas replied:
Fee tokens are not used in the off-ramp so we do not need it, it will be included in the `EVM2AnyMessage` instead

RayXpub commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -200,4 +240,26 @@ library Internal {
+
+  /// @notice Family-agnostic header for OnRamp & OffRamp messages
+  // TODO: revisit if destChainSelector is required (likely sufficient to have it implicitly in the commit roots)
+  struct RampMessageHeader {
```
q: why do we need a separate struct for this?

elatoskinas replied:
The header will be the same across EVM2Any + Any2EVM messages for both the OnRamps and OffRamps, so I've followed the proposal to combine this. However I think we need to revisit the gas cost & contract size implications of this

connorwstein replied:
yeah I just included it to avoid duplication. Totally fine to duplicate w a comment if its better gas wise

elatoskinas replied:
Gas savings seem insignificant (<400 per execute), the contract size as well (<0.03KB) - I think we can keep it for de-duplication and readability

elatoskinas committed:
feat: remove messageId equality check with message hash

elatoskinas committed:
fix: add missing fields to message hash

elatoskinas committed:
feat: implement dest chain selector validation

elatoskinas committed:
Merge branch 'ccip-develop' into feat/any2evm-offramp-structs

elatoskinas committed:
chore: re-gen wrappers and snapshots

elatoskinas committed:
feat: add any2evm hash function to MessageHasher

makramkd commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -200,4 +246,29 @@ library Internal {
+  /// The messageId is not expected to match hash(message), since it may originate from another ramp family
+  // TODO: revisit if destChainSelector is required (likely sufficient to have it implicitly in the commit roots)
+  struct RampMessageHeader {
+    bytes32 messageId; // Unique identifier for the message, generated with the source chain's encoding sdcheme (i.e. not necessarily abi.encoded)
```
```suggestion
    bytes32 messageId; // Unique identifier for the message, generated with the source chain's encoding scheme (i.e. not necessarily abi.encoded)
```

makramkd commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -200,4 +246,29 @@ library Internal {
     Execution
   }
+
+  /// @notice Family-agnostic header for OnRamp & OffRamp messages.
```
Just to make sure I understand, what do we mean by family-agnostic header here? As in, all ramp headers must have this data across all families, but it will get encoded differently per-family?

elatoskinas replied:
> As in, all ramp headers must have this data across all families, but it will get encoded differently per-family?

Yup exactly - as in this `RampMessageHeader` is essentially an `Any2AnyMessageHeader`, it has no coupling to EVM or any other family

makramkd commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -200,4 +246,29 @@ library Internal {
+    uint64 sourceChainSelector; // ───────╮ the chain selector of the source chain, note: not chainId
+    uint64 destChainSelector; //          | the chain selector of the destination chain, note: not chainId
+    uint64 sequenceNumber; //             │ sequence number, not unique across lanes
+    uint64 nonce; // ─────────────────────╯ nonce for this lane for this sender, not unique across senders/lanes
```
OOO behavior still based off of this nonce right, if == 0 then execution is OOO?

makramkd replied:
Ah yeah just saw the ramp code, seems like it.

elatoskinas replied:
Yeah - the behaviour stays the same, the unit tests were preserved from the single-lane ramp version

elatoskinas committed:
fix: typo

makramkd commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -71,7 +71,7 @@ library Internal {
   struct ExecutionReportSingleChain {
     uint64 sourceChainSelector; // Source chain selector for which the report is submitted
-    EVM2EVMMessage[] messages;
+    Any2EVMRampMessage[] messages;
```
@winder just confirming, does this affect our chainlink-common types or is it just the report encoding thats affected?

winder replied:
Yes, this probably effects our types. We'll need to know the exact format of EVM2EVMMessage / Any2EVMRampMessage, and make some sort of provider to make sure we get the right thing.

Hopefully they're close enough that the differences can be baked into the encoding codec. If not, I suppose we'll need to make another provider.

makramkd commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -130,6 +132,19 @@ library Internal {
     });
   }
 
+  function _toAny2EVMMessage(
+    Any2EVMRampMessage memory original,
+    Client.EVMTokenAmount[] memory destTokenAmounts
+  ) internal pure returns (Client.Any2EVMMessage memory message) {
```
Where is this going to be used? Can't see usages in this PR

elatoskinas replied:
This is an `Any2EVM` overload for this original function:

```solidity
  function _toAny2EVMMessage(
    EVM2EVMMessage memory original,
    Client.EVMTokenAmount[] memory destTokenAmounts
  ) internal pure returns (Client.Any2EVMMessage memory message) {
```

Used when forwarding the message to the router from the offramp: https://github.com/smartcontractkit/ccip/blob/ccip-develop/contracts/src/v0.8/ccip/offRamp/EVM2EVMMultiOffRamp.sol#L507

makramkd commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -158,6 +173,37 @@ library Internal {
+    // This hashing scheme is also used by RMN. If changing it, please notify the RMN maintainers.
+    return keccak256(
+      abi.encode(
+        MerkleMultiProof.LEAF_DOMAIN_SEPARATOR,
```
OK to use the same leaf domain separator as before? Metadata hash should cover it right?

elatoskinas replied:
Yeah leaf domain separator is okay. For EVM2EVM -> Any2EVM, we distinguish it with the new `ANY_2_EVM_MESSAGE_HASH` , which is different from the previous `EVM_2_EVM_MESSAGE_HASH` - so the implicit metadata hash covers it

makramkd commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -200,4 +246,29 @@ library Internal {
+    bytes sender; // sender address on the source chain
+    bytes data; // arbitrary data payload supplied by the message sender
+    address receiver; // receiver address on the destination chain
+    uint256 gasLimit; // user supplied maximum gas amount available for dest chain execution
```
How are we handling potentially new fields in this structure? Will it necessarily be an ABI breakage? Suppose there's some other field `uint256 newField` do we come here and create a new hash prefix, hash function, etc.?

elatoskinas replied:
Yes - it would be a breaking change, especially for off-chain reporting - since they would now have to encode messages in a different format. And the hashing would definitely have to change as well (basically each field in the message has to be hashed in, otherwise `manuallyExecute` can bypass and pass in incorrect data)

We should be okay with breaking changes here though, because I don't expect this format to change often (most changes will be in new family chains)

makramkd commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -158,6 +173,37 @@ library Internal {
+        ),
+        keccak256(
+          abi.encode(
+            original.header.messageId,
+            original.sender,
+            original.receiver,
+            original.header.sequenceNumber,
+            original.gasLimit,
+            original.header.nonce
```
Does order of these fields matter here or can you e.g put all the header fields first then the non-header fields next?

elatoskinas replied:
Since we're shifting from EVM2EVM -> family-agnostic messages, we can safely re-order, especially since we have a new hash prefix. For now I just preserved the original order from EVM2EVM:

```solidity
        keccak256(
          abi.encode(
            original.sender,
            original.receiver,
            original.sequenceNumber,
            original.gasLimit,
            original.strict,
            original.nonce,
            original.feeToken,
            original.feeTokenAmount
          )
        )
```

elatoskinas replied:
@gtklocker preference was to keep the order as close to EVM2EVM as possible

elatoskinas committed:
Merge branch 'ccip-develop' into feat/any2evm-offramp-structs

elatoskinas committed:
chore: regen wrappers

makramkd committed:
plugins/ccipevm: impl any2evm message hashing (#1142)

## Motivation

Recent PR #1117 implemented Any2EVM message hashing, so we need to
update the MessageHasher for EVM to match the new onchain
implementation.

## Solution

Update the message hasher to match the `_hash(Any2EVMRampMessage)`
implementation in Internal.sol.

elatoskinas committed:
Merge branch 'ccip-develop' into feat/any2evm-offramp-structs

elatoskinas committed:
fix: lint issue

asoliman92 commented on contracts/src/v0.8/ccip/offRamp/EVM2EVMMultiOffRamp.sol:
```diff
@@ -728,21 +725,20 @@ contract EVM2EVMMultiOffRamp is ITypeAndVersion, MultiOCR3Base {
-        currentConfig.onRamp = sourceConfigUpdate.onRamp;
-        currentConfig.minSeqNr = 1;
+      if (currentOnRamp.length == 0) {
+        if (newOnRamp.length == 0) {
```
I think will be more efficient to check the newOnRamp first. Most probably the currentOnRamp will already have an address.

elatoskinas replied:
Length checks are cheap so it wouldn't be substantial, but in this case this order makes sense because the `onRamp` can only be set if it was never set before (we treat it as immutable). `newOnRamp.length == 0` checks are not required if the currentOnRamp is checked because we're checking equality directly

RayXpub commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -130,6 +132,19 @@ library Internal {
     });
   }
 
+  function _toAny2EVMMessage(
```
nit: we should add some comments to explain why we need another `_toAny2EVMMessage` i.e tag the older one as "necessary for backward compatibility"

elatoskinas replied:
Looks like inlining this saves a good amount of contract size (~0.07KB) - https://github.com/smartcontractkit/ccip/pull/1147/files#diff-8deef24f4fc1cc3e29f8e8c9eff38771d3b66607e4f182e6c8d79feae4888935L120

So I'll leave this one as-is and inline the other one

RayXpub commented on contracts/src/v0.8/ccip/libraries/Internal.sol:
```diff
@@ -158,6 +173,37 @@ library Internal {
 
+  bytes32 internal constant ANY_2_EVM_MESSAGE_HASH = keccak256("Any2EVMMessageHashV1");
+
+  function _hash(Any2EVMRampMessage memory original, bytes memory onRamp) internal pure returns (bytes32) {
```
nit: same as https://github.com/smartcontractkit/ccip/pull/1117/files#r1666920813

elatoskinas replied:
Removed - now inlined

elatoskinas replied:
Added comments to both functions

elatoskinas committed:
refactor: inline any2evmmessage conversion

elatoskinas committed:
docs: hash functions for ramp messages

elatoskinas committed:
Merge branch 'ccip-develop' into feat/any2evm-offramp-structs
